File: backend/tests/trombi.test.js
const request = require('supertest');
const fs = require('fs');
const os = require('os');
const path = require('path');

jest.mock('../src/config/database', () => ({
  query: jest.fn(),
  connect: jest.fn(),
  on: jest.fn(),
}));

// We mock pdfService so tests don't need a real PDF renderer
// Note: jest.mock factories are hoisted, so use require() instead of top-level variables
jest.mock('../src/services/pdfService', () => ({
  generatePDF: jest.fn().mockResolvedValue(
    require('path').join(require('os').tmpdir(), 'trombi-test.pdf')
  ),
}));

const FAKE_PDF = path.join(os.tmpdir(), 'trombi-test.pdf');

const app = require('../src/app');
const pool = require('../src/config/database');

const MOCK_STUDENTS = [
  { id: 1, first_name: 'Jean', last_name: 'Dupont', email: 'jean@school.fr', class_id: 1, photo_url: null },
];

beforeAll(() => {
  // Create a fake PDF file for streaming tests
  fs.writeFileSync(FAKE_PDF, '%PDF-1.4\n%%EOF\n');
});

afterAll(() => {
  if (fs.existsSync(FAKE_PDF)) fs.unlinkSync(FAKE_PDF);
});

describe('Trombi API', () => {
  beforeEach(() => jest.clearAllMocks());

  // ── HTML format ───────────────────────────────────────
  describe('GET /api/trombi?format=html', () => {
    it('returns status 200 and an HTML document', async () => {
      pool.query
        .mockResolvedValueOnce({ rows: MOCK_STUDENTS }) // Student.findAll
        .mockResolvedValueOnce({ rows: [{ id: 1 }] });  // Export.create

      const res = await request(app).get('/api/trombi?format=html');
      expect(res.status).toBe(200);
      expect(res.headers['content-type']).toMatch(/text\/html/);
      expect(res.text).toContain('Trombinoscope');
      expect(res.text).toContain('Jean');
      expect(res.text).toContain('RGPD');
    });

    it('includes class label when class_id is provided', async () => {
      pool.query
        .mockResolvedValueOnce({ rows: MOCK_STUDENTS })               // Student.findAll
        .mockResolvedValueOnce({ rows: [{ id: 1, label: '3A' }] })   // Class.findById
        .mockResolvedValueOnce({ rows: [{ id: 1 }] });                // Export.create

      const res = await request(app).get('/api/trombi?format=html&class_id=1');
      expect(res.status).toBe(200);
      expect(res.text).toContain('3A');
    });

    it('returns 404 when no students are found', async () => {
      pool.query.mockResolvedValue({ rows: [] });

      const res = await request(app).get('/api/trombi?format=html');
      expect(res.status).toBe(404);
    });

    it('returns 400 for unsupported format', async () => {
      const res = await request(app).get('/api/trombi?format=docx');
      expect(res.status).toBe(400);
    });
  });

  // ── PDF format ────────────────────────────────────────
  describe('GET /api/trombi?format=pdf', () => {
    it('returns a non-empty PDF file', async () => {
      pool.query
        .mockResolvedValueOnce({ rows: MOCK_STUDENTS }) // Student.findAll
        .mockResolvedValueOnce({ rows: [{ id: 1 }] });  // Export.create

      const res = await request(app)
        .get('/api/trombi?format=pdf')
        .buffer(true)
        .parse((res, callback) => {
          const chunks = [];
          res.on('data', (c) => chunks.push(c));
          res.on('end', () => callback(null, Buffer.concat(chunks)));
        });

      expect(res.status).toBe(200);
      expect(res.headers['content-type']).toMatch(/application\/pdf/);
      expect(res.body.length).toBeGreaterThan(0);
    });
  });
});

File: backend/tests/trombi.test.py
import os
import tempfile
import pytest
from unittest.mock import patch
from fastapi.testclient import TestClient

from src.main import app
from src.middlewares.auth import get_current_user

MOCK_USER = {"id": 1, "username": "testuser", "email": "test@school.fr"}
MOCK_STUDENTS = [
    {"id": 1, "first_name": "Jean", "last_name": "Dupont", "email": "jean@school.fr", "class_id": 1, "photo_url": None},
]


@pytest.fixture(autouse=True)
def override_auth():
    app.dependency_overrides[get_current_user] = lambda: MOCK_USER
    yield
    app.dependency_overrides.pop(get_current_user, None)


client = TestClient(app)


class TestTrombiHTML:
    def test_returns_200_and_html_document(self):
        with patch("src.routers.trombi.StudentModel.find_all") as mock_students, \
             patch("src.routers.trombi.ExportModel.create"):
            mock_students.return_value = MOCK_STUDENTS
            res = client.get("/api/trombi/?format=html")
        assert res.status_code == 200
        assert "text/html" in res.headers["content-type"]
        assert "Trombinoscope" in res.text
        assert "Jean" in res.text
        assert "RGPD" in res.text

    def test_includes_class_label_when_class_id_provided(self):
        with patch("src.routers.trombi.StudentModel.find_all") as mock_students, \
             patch("src.routers.trombi.ClassModel.find_by_id") as mock_class, \
             patch("src.routers.trombi.ExportModel.create"):
            mock_students.return_value = MOCK_STUDENTS
            mock_class.return_value = {"id": 1, "label": "3A"}
            res = client.get("/api/trombi/?format=html&class_id=1")
        assert res.status_code == 200
        assert "3A" in res.text

    def test_returns_404_when_no_students_found(self):
        with patch("src.routers.trombi.StudentModel.find_all") as mock_students:
            mock_students.return_value = []
            res = client.get("/api/trombi/?format=html")
        assert res.status_code == 404

    def test_returns_400_for_unsupported_format(self):
        res = client.get("/api/trombi/?format=docx")
        assert res.status_code == 400


class TestTrombiPDF:
    def test_returns_pdf_file(self):
        with tempfile.NamedTemporaryFile(suffix=".pdf", delete=False) as f:
            f.write(b"%PDF-1.4\n%%EOF\n")
            fake_pdf_path = f.name
        try:
            with patch("src.routers.trombi.StudentModel.find_all") as mock_students, \
                 patch("src.routers.trombi.generate_pdf") as mock_pdf, \
                 patch("src.routers.trombi.ExportModel.create"):
                mock_students.return_value = MOCK_STUDENTS
                mock_pdf.return_value = fake_pdf_path
                res = client.get("/api/trombi/?format=pdf")
            assert res.status_code == 200
            assert "application/pdf" in res.headers["content-type"]
            assert len(res.content) > 0
        finally:
            if os.path.exists(fake_pdf_path):
                os.unlink(fake_pdf_path)
